### PR TITLE
Implement cloudprovider.InstancesV2 interface

### LINKED ISSUE #344: Add support for InstancesV2 interface
Refer PR:

* https://github.com/kubernetes/kubernetes/pull/92367 (merged)
* https://github.com/kubernetes/kubernetes/pull/91319 (merged)

After these PR gets merged, we could add InstancesV2 support for Cloud Provider Azure. The codes should be implemented in k/k and vendor into cloud-provider-azure repo.

### PR DESCRIPTION
**Pull request by nilo19**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake
/kind feature

**What this PR does / why we need it**:
Implement cloudprovider.InstancesV2 interface

**Which issue(s) this PR fixes**:

Fixes #344

**Special notes for your reviewer**:

**Release note**:
```
Implement cloudprovider.InstancesV2 interface
```

### CODE REVIEW

File: pkg/provider/azure_instances.go [hunk 1]
```diff
@@ -304,6 +340,15 @@ func (az *Cloud) InstanceID(ctx context.Context, name types.NodeName) (string, e
 	return az.VMSet.GetInstanceIDByNodeName(nodeName)
 }
 
+func (az *Cloud) getNodeProviderIDByNodeName(ctx context.Context, name types.NodeName) (string, error) {
+	providerID, err := cloudprovider.GetInstanceProviderID(ctx, az, name)
```
Comment (feiskyer) [inline pkg/provider/azure_instances.go:344]: use `az.InstanceID()` instead?

File: pkg/provider/azure_instances.go [hunk 2]
```diff
@@ -207,6 +207,24 @@ func (az *Cloud) InstanceExistsByProviderID(ctx context.Context, providerID stri
...
+func (az *Cloud) InstanceExists(ctx context.Context, node *v1.Node) (bool, error) {
+	if node == nil {
+		return false, nil
+	}
+	providerID := node.Spec.ProviderID
+	if providerID == "" {
+		var err error
+		providerID, err = az.getNodeProviderIDByNodeName(ctx, types.NodeName(node.Name))
+		if err != nil {
+			return false, err
```
Comment (feiskyer) [inline pkg/provider/azure_instances.go:221]: Please add some error logs in the exposed functions.

File: pkg/provider/azure_instances.go (hunk 1 above)
Comment (nilo19) [reply pkg/provider/azure_instances.go:344]: The `cloudprovider.GetInstanceProviderID` put a prefix `azure://` onto the `az.InstanceID`, which is the standard provider ID format. Some regexps need this such as https://github.com/kubernetes-sigs/cloud-provider-azure/blob/d12fc98b424d6a33e546679577825daccbc45165/pkg/provider/azure_vmss.go#L54

Comment (feiskyer) [reply pkg/provider/azure_instances.go:346]: OK, then it makes sense here.

**Review by feiskyer — APPROVED** [review]
/lgtm

File: pkg/provider/azure.go [hunk 3]
```diff
@@ -686,9 +686,8 @@ func (az *Cloud) Instances() (cloudprovider.Instances, bool) {
 }
 
 // InstancesV2 returns an instancesV2 interface. Also returns true if the interface is supported, false otherwise.
-// TODO: implement ONLY for external cloud provider
 func (az *Cloud) InstancesV2() (cloudprovider.InstancesV2, bool) {
-	return nil, false
+	return az, false
```
Comment (andyzhangx) [inline pkg/provider/azure.go:690]: should it be `az, true` indicates it supports InstancesV2? and when are we switching to use InstancesV2, is there e2e test to cover this?

**Comment by andyzhangx** [comment]
/hold
hold a moment

File: pkg/provider/azure.go (hunk 3 above)
Comment (nilo19) [reply pkg/provider/azure.go:690]: Yeah, that would be `true` here. The `InstancesV2` would be enabled once this is merged.

Comment (nilo19) [reply pkg/provider/azure.go:690]: Tested manually to make sure the information on the node is good.

**Review by andyzhangx — APPROVED** [review]
/hold cancel
/lgtm

**Pull request merged**